libs/execution_context/ebpf_maps.c

```diff
@@ -872,40 +952,193 @@ _delete_lru_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
+{
+    if (map->hot_list_size <= map->hot_list_limit) {
+        return;
+    }
+
+    ebpf_list_entry_t* list_entry = map->hot_list.Flink;
+    ebpf_list_remove_entry(&map->hot_list);
+    ebpf_list_append_tail_list(&map->cold_list, list_entry);
+
+    ebpf_list_initialize(&map->hot_list);
+    map->hot_list_size = 0;
+    map->current_generation++;
+}
+
 static ebpf_result_t
-_update_key_history(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key, _In_opt_ uint8_t* value)
+_update_key_history(
+    _In_ ebpf_core_map_t* map, _In_ const uint8_t* key, _In_opt_ uint8_t* value, ebpf_lru_key_operation_t operation)
 {
+    // Fast path:
+    // If this map doesn't track key history, return success.
     if (!(ebpf_map_metadata_tables[map->ebpf_map_definition.type].key_history)) {
         return EBPF_SUCCESS;
     }
 
+    // If value was not provided, attempt to find it.
     if (value == NULL) {
         ebpf_result_t find_result = ebpf_hash_table_find((ebpf_hash_table_t*)map->data, key, &value);
         if (find_result != EBPF_SUCCESS) {
             return find_result;
         }
     }
+
+    // If the value is not found, return success as it has already been deleted.
+
+    // If the value is found, update the key history.
     if (value) {
-        uint64_t* key_age = (uint64_t*)_get_value_supplement(map, value);
-        *key_age = ebpf_query_time_since_boot(false);
+        ebpf_lru_entry_t* entry = (ebpf_lru_entry_t*)_get_value_supplement(map, value);
+        ebpf_core_lru_map_t* lru_map = EBPF_FROM_FIELD(ebpf_core_lru_map_t, core_map, map);
+
+        ebpf_lru_key_state_t key_state = _get_key_state(lru_map, entry);
+
+        // Fast path:
+        // If the key is already in the hot list, no need to update the key history.
+        if (key_state == EBPF_LRU_KEY_HOT && operation == EBPF_LRU_KEY_OPERATION_UPDATE) {
+            // Key is already in the hot list. No need to update the key history.
+            return EBPF_SUCCESS;
+        }
+
+        ebpf_lock_state_t state = ebpf_lock_lock(&lru_map->lock);
+
+        // Code needs to handle all possible combinations of key_state and operation.
+        switch (operation) {
+        case EBPF_LRU_KEY_OPERATION_INSERT:
+            _initialize_lru_entry(lru_map, entry, key);
+            break;
+        case EBPF_LRU_KEY_OPERATION_UPDATE:
+            _insert_into_hot_list(lru_map, entry);
+            break;
+        case EBPF_LRU_KEY_OPERATION_DELETE:
+            _uninitialize_lru_entry(lru_map, entry);
+            break;
+        }
+
+        // If the hot list is full, move the entire hot list to the cold list, and start a new generation.
+        _merge_hot_into_cold_list_if_needed(lru_map);
+
+        ebpf_lock_unlock(&lru_map->lock, state);
     }
+
     return EBPF_SUCCESS;
 }
 
-static bool
+static ebpf_result_t
+_delete_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key);
+
+static ebpf_result_t
 _reap_oldest_map_entry(_In_ ebpf_core_map_t* map)
```

Comment:
```suggestion
_reap_oldest_map_entry(_Inout_ ebpf_core_map_t* map)
```

Reply:
Filed https://github.com/microsoft/ebpf-for-windows/issues/1746 as there are a lot of cases where we should be using _Inout_ or _In_ const.